WalletWasabi/Tor/Socks5/TorSocks5ClientFactory.cs

```diff
@@ -0,0 +1,408 @@
+
+		private EndPoint TorSocks5EndPoint { get; }
+
+		/// <summary>
+		/// Checks whether communication can be established with Tor over <see cref="TorSocks5EndPoint"/> endpoint.
+		/// </summary>
+		/// <returns></returns>
+		public async Task<bool> IsTorRunningAsync()
+		{
+			try
+			{
+				// Internal TCP client may close, so we need a new instance here.
+				using var tcpClient = new TcpClient(TorSocks5EndPoint.AddressFamily);
+				await ConnectAsync(tcpClient).ConfigureAwait(false);
+				await HandshakeAsync(tcpClient, isolateStream: false).ConfigureAwait(false);
+
+				return true;
+			}
+			catch (Exception e) when (e is TorConnectionException or TorAuthenticationException)
+			{
+				return false;
+			}
+		}
+
+		/// <summary>
+		/// Creates a new connected TCP client connected to Tor SOCKS5 endpoint.
+		/// </summary>
+		/// <inheritdoc cref="EstablishConnectionAsync(string, int, bool, bool, CancellationToken)"/>
+		public async Task<IPoolItem> EstablishConnectionAsync(HttpRequestMessage request, bool isolateStream, CancellationToken token = default)
+		{
+			bool useSsl = request.RequestUri!.Scheme == Uri.UriSchemeHttps;
+			string host = request.RequestUri.DnsSafeHost;
+			bool allowRecycling = !useSsl && !isolateStream;
+			int port = request.RequestUri!.Port;
+
+			TorConnection newClient = await EstablishConnectionAsync(host, port, useSsl, isolateStream, token).ConfigureAwait(false);
+			return new TorPoolItem(newClient, allowRecycling);
+		}
+
+		/// <summary>
+		/// Creates a new connected TCP client connected to Tor SOCKS5 endpoint.
+		/// </summary>
+		/// <param name="host">Tor SOCKS5 host.</param>
+		/// <param name="port">Tor SOCKS5 port.</param>
+		/// <param name="useSsl">Whether to use SSL to send the HTTP request over Tor.</param>
+		/// <param name="isolateStream"><c>true</c> if a new Tor circuit is required for this HTTP request.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel the asynchronous operation.</param>
+		/// <returns>New <see cref="TorConnection"/> instance.</returns>
+		/// <exception cref="TorConnectionException">When <see cref="ConnectAsync(TcpClient, CancellationToken)"/> fails.</exception>
+		public async Task<TorConnection> EstablishConnectionAsync(string host, int port, bool useSsl, bool isolateStream, CancellationToken cancellationToken = default)
+		{
+			TcpClient? tcpClient = null;
+			Stream? transportStream = null;
+
+			try
+			{
+				tcpClient = new TcpClient(TorSocks5EndPoint.AddressFamily);
+
+				transportStream = await ConnectAsync(tcpClient, cancellationToken).ConfigureAwait(false);
+				await HandshakeAsync(tcpClient, isolateStream, cancellationToken).ConfigureAwait(false);
+				await ConnectToDestinationAsync(tcpClient, host, port, cancellationToken).ConfigureAwait(false);
+
+				if (useSsl)
+				{
+					transportStream = await UpgradeToSslAsync(tcpClient, host).ConfigureAwait(false);
+				}
+
+				var result = new TorConnection(tcpClient, transportStream);
```

Comment:
They won't be disposed because I set `transportStream` to `null` immediately after this line.

It works like this: You want to create `tcpClient` and `transportStream` but any time in the process of construction of these objects there may be an exception. If an exception is thrown, both `tcpClient` and `transportStream` need to be disposed. If no exception is thrown in `try` block then nothing is disposed.

Reply:
Oh I see, it is a bit tricky.